packages/browser-repl/src/components/shell.tsx

```diff
@@ -153,94 +163,123 @@ const normalizeInitialEvaluate = (initialEvaluate: string | string[]) => {
-    output: this.props.initialOutput.slice(-this.props.maxOutputLength),
-    history: this.props.initialHistory.slice(0, this.props.maxHistoryLength),
-    passwordPrompt: '',
-    shellPrompt: '>',
-  };
-
-  componentDidMount(): void {
-    // Store the intial prop value on mount so that we're not using potentially
-    // updated one when actually running the lines
-    let evalLines: string[] = [];
-    if (this.props.initialEvaluate) {
-      evalLines = normalizeInitialEvaluate(this.props.initialEvaluate);
-    }
-    this.scrollToBottom();
-    void this.updateShellPrompt().then(async () => {
-      for (const input of evalLines) {
-        await this.onInput(input);
-      }
-    });
-  }
+const capLengthEnd = (elements: unknown[], maxLength: number) => {
+  elements.splice(0, elements.length - maxLength);
+};
 
-  componentDidUpdate(): void {
-    this.scrollToBottom();
-  }
+const capLengthStart = (elements: unknown[], maxLength: number) => {
+  elements.splice(maxLength);
+};
 
-  private evaluate = async (code: string): Promise<ShellOutputEntry> => {
-    let outputLine: ShellOutputEntry;
+// eslint-disable-next-line react/display-name
+export const Shell = ({
+  runtime,
+  className,
+  redactInfo,
+  maxOutputLength = 1000,
+  maxHistoryLength = 1000,
+  onInputChanged,
+  onOutputChanged,
+  onHistoryChanged,
+  onOperationStarted,
+  onOperationEnd,
+  initialEvaluate,
+  inputText,
+  output,
+  history,
+  isOperationInProgress = false,
+  onEditorChanged,
+}: ShellProps) => {
+  const darkMode = useDarkMode();
 
-    try {
-      this.props.onOperationStarted();
-
-      this.props.runtime.setEvaluationListener(this);
-      const result = await this.props.runtime.evaluate(code);
-      outputLine = {
-        format: 'output',
-        type: result.type,
-        value: result.printable,
-      };
-    } catch (error) {
-      outputLine = {
-        format: 'error',
-        value: error,
-      };
-    } finally {
-      await this.updateShellPrompt();
-      this.props.onOperationEnd();
-    }
+  const [editor, setEditor] = useState<EditorRef | null>(null);
+  const [passwordPrompt, setPasswordPrompt] = useState('');
+  const [shellPrompt, setShellPrompt] = useState('>');
+  const [shellInputElement, setShellInputElement] =
+    useState<HTMLElement | null>(null);
+  const [onFinishPasswordPrompt, setOnFinishPasswordPrompt] = useState<
+    () => (result: string) => void
+  >(() => noop);
+  const [onCancelPasswordPrompt, setOnCancelPasswordPrompt] = useState<
+    () => () => void
+  >(() => noop);
+
+  const focusEditor = useCallback(() => {
+    editor?.focus();
+  }, [editor]);
+
+  const editorChanged = useCallback(
+    (editor: EditorRef | null) => {
+      setEditor(editor);
+      onEditorChanged?.(editor);
+    },
+    [onEditorChanged]
+  );
+
+  const listener = useMemo(() => {
+    const evaluationListener: RuntimeEvaluationListener = {
```

Comment:
nit: could probably just do this if we're going for type safety

```suggestion
  const listener = useMemo<RuntimeEvaluationListener>(() => {
    return {
```